### ISSUE TITLE
[flang] Misleading parser diagnostic for a non-component statement after CONTAINS in a derived type

### ISSUE THREAD

**Issue by kwyatt-ext** [llvm/llvm-project#215849]
# [flang] Misleading parser diagnostic for a non-component statement after CONTAINS in a derived type

## Summary

When a statement that is not a type-bound procedure binding appears after
`CONTAINS` in a derived-type definition, flang now emits a misleading error that
talks about intrinsic type-spec keywords (`expected 'COMPLEX'`,
`expected 'INTEGER'`, ...) instead of indicating that a type-bound procedure
binding was expected. In some cases (e.g. a misplaced subprogram) it emits an
avalanche of ~20 unrelated "expected '<type-keyword>'" errors.

The programs are all correctly rejected -- this is a **diagnostic-quality
regression**, not a miscompile.

This was introduced by #203379

## Reproducers

`repro_contains.f90` -- a second `CONTAINS`:
```fortran
type :: t
contains
contains
end type
end
```

`repro_import.f90` -- an `IMPORT` after `CONTAINS`:
```fortran
module m
  type :: t
  contains
    import
  end type
end module
```

`repro_sub.f90` -- a misplaced subprogram after `CONTAINS`:
```fortran
module m
  type :: t
  contains
    subroutine s
    end subroutine
  end type
end module
```

## Current (bad) output

`repro_contains.f90`:
```
error: expected 'COMPLEX'
  contains
  ^
  in the context: intrinsic type spec
  in the context: type bound procedure part
```

`repro_import.f90`:
```
error: expected 'INTEGER'
      import
      ^
  in the context: intrinsic type spec
  in the context: type bound procedure part
```

`repro_sub.f90` is the worst case -- a single misplaced `subroutine` produces
~20 errors, cycling through every type-spec keyword before finally listing the
real alternatives:
```
error: expected '__VECTOR_QUAD'   ... in the context: vector type spec
error: expected '__VECTOR_PAIR'   ... in the context: vector type spec
error: expected 'VECTOR'          ... in the context: vector type spec
error: expected 'RECORD'          ... in the context: declaration type spec
error: expected 'CLASS'           ... in the context: declaration type spec
error: expected 'TYPE'            ... in the context: declaration type spec
...
error: expected 'COMPLEX'         ... in the context: intrinsic type spec
error: expected 'REAL'            ... in the context: intrinsic type spec
error: expected 'INTEGER'         ... in the context: intrinsic type spec
error: expected 'FINAL'           ... in the context: FINAL statement
error: expected 'GENERIC'         ... in the context: type bound GENERIC statement
error: expected 'PROCEDURE'       ... in the context: type bound PROCEDURE statement
```

## Expected output

A single, clear diagnostic that a type-bound procedure binding was expected,
e.g.:
```
error: expected a type-bound procedure binding (PROCEDURE, GENERIC, or FINAL) after CONTAINS
  in the context: type bound procedure binding
  in the context: type bound procedure part
```

## Root cause

#203379 added `DataComponentDefStmt` as a trailing alternative to the
`type-bound-proc-binding` parser in `flang/lib/Parser/Fortran-parsers.cpp` so
that a genuinely misplaced *component definition* (e.g.
`integer, public :: n(3) = 1`) after `CONTAINS` yields the nice message
`component definition must precede CONTAINS in a derived type`:

```cpp
TYPE_CONTEXT_PARSER("type bound procedure binding"_en_US,
    recovery(
        first(construct<TypeBoundProcBinding>(Parser<TypeBoundProcedureStmt>{}),
            construct<TypeBoundProcBinding>(Parser<TypeBoundGenericStmt>{}),
            construct<TypeBoundProcBinding>(Parser<FinalProcedureStmt>{}),
            Parser<DataComponentDefStmt>{} >>            // <-- added by #203379
                fail<TypeBoundProcBinding>(
                    "component definition must precede CONTAINS in a derived type"_err_en_US)),
        construct<TypeBoundProcBinding>(
            !"END"_tok >> SkipTo<'\n'>{} >> construct<ErrorRecovery>())))
```

The intended `fail<>()` message only fires when `DataComponentDefStmt` **fully**
matches. For any *other* invalid statement after `CONTAINS` (a second
`CONTAINS`, `IMPORT`, a misplaced subprogram, ...), `DataComponentDefStmt`
begins by attempting `declarationTypeSpec` -> intrinsic-type-spec, which is a
`first()` over `INTEGER / REAL / ... / COMPLEX / ...`. That attempt fails, but
its internal "expected '<type-keyword>'" messages leak out of the alternative.
In `CombineFailedParses` the new alternative's messages then displace the
previous `FINAL/GENERIC/PROCEDURE` recovery messages (and, depending on how far
the partial parse advances, either surface a single arbitrary type keyword or
the whole merged set of them). The net effect is a strictly worse diagnostic for
every non-component statement after `CONTAINS`.

The regression test added by that PR (`flang/test/Parser/recovery09.f90`) only
exercises the valid-component case, so the degraded path for non-component
statements was not caught.

Assisted-by: AI

**Comment by kwyatt-ext** [comment]
@mleair @eugeneepshteyn @akuhlens 
I have a patch ready that will produce the "Expected Output" listed above.  I just wanted to ping the author and reviewers of the original PR before jumping in.  I'll put a PR in shortly if there are no objections.

**Comment by mleair** [comment]
Your proposal seems fine to me.

**Comment by kwyatt-ext** [comment]
Sounds good.  I'll push the PR.  Thank you.
